apps/lib/app_provider.c

```diff
@@ -65,6 +66,79 @@ static int opt_provider_path(const char *path)
     return OSSL_PROVIDER_set_default_search_path(app_get0_libctx(), path);
 }
 
+struct prov_param_st {
+    char *name;
+    char *key;
+    char *val;
+    int found;
+};
+
+static int set_prov_param(OSSL_PROVIDER *prov, void *vp)
+{
+    struct prov_param_st *p = (struct prov_param_st *)vp;
+
+    if (p->name != NULL
+        && strcasecmp(OSSL_PROVIDER_get0_name(prov), p->name) != 0)
```

Comment:
You need to always use `OPENSSL_strcasecmp`.

Also this case-insensitive matching is inconsistent with the fact, that you can load two different providers differing with just case such as `FIPS` vs `fips` providers. Of course the `provider` property is also case insensitive, so it will not be possible to distinguish them just by a `provider=FIPS` property query but perhaps the hypothetical `FIPS` provider would have another property in its algorithms to distinguish it from our `fips` provider.

Of course maybe the best idea would be to make the provider name matching in ossl_provider_cmp() in provider_core.c case-insensitive as well. It would be a breaking change but perhaps acceptable for a minor release with an appropriate CHANGES.md entry.

Reply:
> Also this case-insensitive matching is inconsistent with the fact, that you can load two different providers differing with just case such as FIPS vs fips providers

There is a lot of inconsistency here. In general I think case insensitive is probably the better route. I don't think we really need to worry too much about the case where there are two providers with the same name differing only in case (in that case, we would match both....oh well, we'll live with it).

Reply:
Yeah, I do not insist on having this case sensitive match. But strcasecmp -> OPENSSL_strcasecmp() is necessary!

Reply:
The reason provider=X works for case insensitive handling (which is where I think you'll find is the most common user use of interaction with providers) is because the property handling logic folds all non-quoted string values to lowercase explicitly [here](https://github.com/openssl/openssl/blob/master/crypto/property/property_parse.c#L236). 

This has a completely separate issue - if you elect to name a provider using mixed case then provider=X will not match which I would argue is very much non-expected behaviour. If you use provider="X" you get a case sensitive comparison.

On that basis I would back this change out and tackle provider name matching consistency in a separate PR.

Reply:
OK, backing it out. Simpler than switching to `OPENSSL_strcasetmp`. :-)